hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java

```diff
@@ -13,23 +13,31 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.hedera.node.app.spi.meta;
-
-import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_PAYER_ACCOUNT_ID;
-import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+package com.hedera.node.app;
 
 import com.hedera.node.app.service.token.impl.AccountStore;
 import com.hedera.node.app.spi.key.HederaKey;
+import com.hedera.node.app.spi.meta.TransactionMetadata;
 import com.hederahashgraph.api.proto.java.AccountID;
 import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
 import com.hederahashgraph.api.proto.java.TransactionBody;
+
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_PAYER_ACCOUNT_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+
 /**
  * Metadata collected when transactions are handled as part of "pre-handle" needed for signature
  * verification. This class may have subclasses in the future.
+ *
+ * <p>NOTE: This class shouldn't exist here, and is something of a puzzle. We cannot add it to
+ * SPI, because it includes a dependency on AccountStore. But we also cannot put it in the
+ * app module, because doing so would cause service modules to have a circular dependency
+ * on the app module. Maybe we need some kind of base module from which services can extend
+ * and put it there?
```

Comment:
Is `AccountStore` not a 'service' in this context? I thought `AccountStore` would have a corresponding service interface in the SPI; this class would then go in the app module with a dependency not on `AccountStore`, but on `AccountStore`'s extracted service interface

Reply:
All stores are just implementation details and are responsible for hiding all the Merkleness. We will not be having an interface in SPI for it.

Reply:
Right now `AccountStore` is the implementation, yes, _but it is also serving as the interface_. If we want `AccountStore` to _only_ be an implementation detail, we need to figure out what that interface is.

Reply:
Maybe instead of taking an `AccountStore` it should take an `AccountService`? Not sure. The reason we must keep the stores as implementation details is because if we allow one service to have access to another service's data store, then we provide a back door for one service to mutates the state of another without going through "the front door".

Reply:
After looking at this class in more depth, I feel like it doesn't need a dependency on the `AccountStore` class at all. The only place the `AccountStore` is used is during the constructor call (see `addPayerKey(payer)` on line 57). The rest of the class is a pojo. Makes more sense to me to take `AccountStore` out of this class and pass in the data that the pojo needs in order to store its other data. 

@rbair23 this class feels a bit like what you're talking about, 'going through a back door.' No data is changed here, but this code does utilize the account storage mechanism to query for information. If instead the stores are gated by services (and thus interfaces) that have explicit operations on those services, we can truly encapsulate the stores just like we would encapsulate a more traditional persistence object (DAO, Spring repository, etc.) in a layered application. Of course we don't have to do this--I know some people hate the boilerplate of having an interface on top of persistence classes–-but it feels like a much cleaner separation of responsibilities, and a more controlled approach that makes persistence queries and mutations more explicit.

Reply:
Yes exactly. If you think of a traditional micro-services application, and each of our 'services' as a micro-service, then you get the same idea. A microservice doesn't give another microservice direct access to its data store (the db). Instead, it has some REST API or gRPC API that the other services have to use to interact with the microservice. The Microservice aggressively defends its state. In the same way, each of our services needs to defend its `Store` from other services and require them to go through a public API, and we have to be very intentional about what public API we expose. If we expose too much, we tightly couple services together.